_posts/2020-03-23-one-click-install.md

```diff
@@ -0,0 +1,33 @@
+---
+layout: blog-detail
+post-type: blog
+by: Sébastien Doeraene
+title: "One-click install for Scala"
+---
+
+Installing Scala has always been a task more challenging than necessary, with the potential to drive away beginners.
+Should I install Scala itself? sbt? Some other build tools? What about a better REPL like Ammonite? Oh and before all that I need to install Java?
+
+To solve this problem, the Scala Center contracted Alexandre Archambault in January 2020 to add a one-click install of Scala to [coursier](https://get-coursier.io/docs/cli-overview).
+Starting today, [the main Scala download page](/download/) gives one-liner command lines incantations (and a one-click executable for Windows) to fully install Scala and all the main related tools.
+For example, on Linux, all we now need is:
+
+```bash
+$ curl -Lo cs https://git.io/coursier-cli-linux && chmod +x cs && ./cs setup
+```
+
+which will install all the following software, if not yet installed:
+
+* a JDK
+* the build tools [sbt](https://www.scala-sbt.org/) and [mill](https://www.lihaoyi.com/mill/)
```

Comment:
Regarding sbt, would this install the same launcher script and launcher JAR as what one would get using downloading the installer or using SDKMAN?

Reply:
This installs a launcher for the [coursier-based sbt launcher](https://github.com/coursier/sbt-launcher), based on [this "app descriptor"](https://github.com/coursier/apps/blob/d21889a73797cbef3d0f8b0742945b054ac58c0c/apps/resources/sbt-launcher.json).

You can generate one locally with `cs bootstrap sbt-launcher -o sbt` for example.

Reply:
I think it would be very misleading to call any alternative packaging of sbt as `sbt` on novice user's PATH since inevitably they will run into some incompatibilities of handling SBT_OPTS environment variable or dependency resolution difference etc, and then whoever in the position of supporting the person in-house, or in open source would have to reverse-engineer "ok. what `sbt` script are you using, and how did it get to your PATH?"

I'm more than happy to use Coursier for all three layers (launcher, plugins, proper build), but I think it should happen in conjunction with the official (preloaded) distribution.

Reply:
Here's an example in Bash script incompatibility (likely parsing of SBT_OPTS):

```
$ ./sbt
[info] Loading settings for project global-plugins from idea.sbt,vimquit.sbt,plugins.sbt,pgp.sbt ...
[info] Loading global plugins from /Users/eed3si9n/.sbt/1.0/plugins
[info] Loading project definition from /Users/eed3si9n/work/sbt-coursier/project/project
[info] Loading settings for project sbt-coursier-build from plugins.sbt ...
[info] Loading project definition from /Users/eed3si9n/work/sbt-coursier/project
[info] Loading settings for project sbt-coursier-root from build.sbt ...
[info] Set current project to sbt-coursier-root (in build file:/Users/eed3si9n/work/sbt-coursier/)
[error] Expected ';'
[error] -Xms2048M
[error]          ^
```

Reply:
The output of `cs bootstrap sbt-launcher -o sbt` (basically, what is installed by `cs setup` or `cs install sbt-launcher`) doesn't read `SBT_OPTS`. Not sure how you can get that error.

If you used the [custom sbt-extras script](https://github.com/coursier/sbt-extras/blob/master/sbt) linked from the [README](https://github.com/coursier/sbt-launcher/blob/master/README.md), it should basically be the same handling as [sbt-extras mainline](https://github.com/paulp/sbt-extras).

Reply:
So there are now 3 flavors of `sbt` script: official, sbt-extra, and coursier/sbt-launcher? If it's not byte-for-byte equivalent to the official distribution, please call it something else on the PATH, like `csbt` to avoid confusion.

In some future timing, like sbt 1.4.x or whatever, the official one can switch over to something that uses Coursier by default, so if someone runs into a bug / incompatibility we can address it together. sbt releases make changes not only to the core artifacts, but also to Bash / Batch files.

For example, `sbt` supports `--supershell=never` as Bash option. Passing that to `csbt` you'd get:

```
[error] Expected ';'
[error] --supershell=never
[error]                   ^
```

I ❤️ the overall initiative to make CLI setup easier, but putting forked implementation on PATH is not helpful.

Reply:
The coursier-based launcher works in a majority of cases, and also allows to be installed just like any other JVM apps here, with no ad-hoc handling. To me, it's fine if there's a discrepancy between the coursier-based launcher and the others for a short amount of time…

Plus it shouldn't differ that much from sbt-extras, apart from the Java options handling (it only accepts those via `JAVA_OPTS`, just like the other apps installed by `cs setup` or `cs install`). And I would expect those discrepancies to shrink even more if the coursier-based launcher were to use native-image - as it would run the `java` command itself, it could handle Java options too for example, kind of like what you described in https://github.com/sbt/sbt/issues/5406#issuecomment-615401865.

Reply:
it came up on Gitter just now that somebody was using the coursier-based launcher and wasn't getting their `.jvmopts`, in case there's any doubt this will bite people in practice

Reply:
ah, the user in question opened a ticket: https://github.com/coursier/sbt-launcher/issues/167

Reply:
FYI, `cs setup` now installs an sbt launcher based on [sbt/launcher](https://github.com/sbt/launcher). If you already installed `sbt` via `cs setup` or `cs install`, remove it and install `sbt` again, like
```text
$ cs uninstall sbt
$ cs install sbt
```

It still doesn't read `.jvmopts` though. It handles a few of the options of sbt-extras via [coursier/sbt-runner](https://github.com/coursier/sbt-runner) (that parses some options upfront, then calls the main class of sbt/launcher).

Reply:
This is great!

Is there a reason `cs update sbt` would not work?

Reply:
Yes, that's because the name of the "app descriptor" changed from [`sbt-launcher`](https://github.com/coursier/apps/blob/8e124db6395253e5ece202d6bf1dd47cae922170/apps/resources/sbt-launcher.json) to [`sbt`](https://github.com/coursier/apps/blob/8e124db6395253e5ece202d6bf1dd47cae922170/apps/resources/sbt.json). The uninstall-then-install steps allow to pull the newer app descriptor.

Just `sbt` now works, as if one uses it with an explicit version, it does correspond to an actual sbt version as expected. This gives `cs launch sbt:1.3.10 -- new scala/scala-seed.g8`. This works because the new app descriptor references `org.scala-sbt:sbt-launch` directly. With the former one, the explicit version would have applied to just coursier/sbt-launcher, [whose versions](https://github.com/coursier/sbt-launcher/releases) are not in sync with sbt. `cs launch sbt-launcher:1.2.21` felt slightly less odd than `cs launch sbt:1.2.21`. That problem went away now.